client/pd_service_discovery.go

```diff
@@ -97,6 +100,234 @@ type ServiceDiscovery interface {
 	AddServiceAddrsSwitchedCallback(callbacks ...func())
 }
 
+// ServiceClient is an interface that defines a set of operations for a raw PD gRPC client to specific PD server.
+type ServiceClient interface {
+	// GetAddress returns the address information of the PD server.
+	GetAddress() string
+	// GetClientConn returns the gRPC connection of the service client
+	GetClientConn() *grpc.ClientConn
+	// BuildGRPCContext builds a context object with a gRPC context.
+	// ctx: the original context object.
+	// mustLeader: whether must send to leader.
+	BuildGRPCContext(ctx context.Context, mustLeader bool) context.Context
+	// IsLeader returns whether the target PD server is leader.
+	IsLeader() bool
+	// Available returns if the network or other availability for the current service client is available.
+	Available() bool
+	// NeedRetry checks if client need to retry based on the PD server error response.
+	// And It will mark the client as unavailable if the pd error shows the follower can't handle request.
+	NeedRetry(*pdpb.Error, error) bool
+}
+
+var _ ServiceClient = (*pdServiceClient)(nil)
+var _ ServiceClient = (*pdServiceAPIClient)(nil)
+
+type pdServiceClient struct {
+	addr       string
+	conn       *grpc.ClientConn
+	isLeader   bool
+	leaderAddr string
+
+	networkFailure atomic.Bool
+}
+
+func newPDServiceClient(addr, leaderAddr string, conn *grpc.ClientConn, isLeader bool) *pdServiceClient {
```

Comment:
Maybe using the interface as the returned value is better.

```suggestion
func newPDServiceClient(addr, leaderAddr string, conn *grpc.ClientConn, isLeader bool) ServiceClient {
```